### ISSUE TITLE
Replace hardcoded regexes for namespaces(e.g.: (?:[Ii]mage|[Ff]ile) ) by others generated from wgNamespaceIds

### ISSUE THREAD

**Issue by he7d3r** [wikimedia-gadgets/twinkle#103]
Currently, there are some instances of the regex "(?:[Ii]mage|[Ff]ile)" and "(?:[Tt]emplate:)" in the code.

It would be better to use all possible aliases of the "File:" namespace ("Template:", respectively) which are in use in a given wiki.

E.g.: On Portuguese Wikipedia, the regex should be "(?:[Ff]icheiro|[Ii]magem|[Aa]rquivo|[Ii]mage|[Ff]ile)". The code below should work on any other wiki as well:

```
var ids = mw.config.get('wgNamespaceIds'),
    aliases = [],
    first;
for( nsName in mw.config.get('wgNamespaceIds') ){
    if ( ids[nsName] === 6 ) {
        first = nsName.substr(0,1);
        aliases.push( '[' + first.toUpperCase() + first.toLowerCase() + ']' + nsName.substr(1) );
    }
}
alert( '(?:' + aliases.join('|') + ')' );
```

**Comment by Amorymeltzer** [comment]
This is necroposting, but, uh, basically none of the above has yet come to pass.

Regardless, a good first step would be removing the hardcoding of `Morebits.wikipedia`.  MediaWiki-1.16 (circa 2011) added `wgNamespaceIds` and `wgFormattedNamespaces`, which should take care of most of the uses of those objects.  `Morebits.wikipedia.namespaces` is basically a carbon copy of `wgFormattedNamespaces` (with the added advantage that the project name for `project` is used).

There are only three uses of `Morebits.wikipedia.namespaces` and `Morebits.wikipedia.namespacesFriendly` in our codebase.  It shouldn't be too difficult to remove them, but I don't know who or what else relies on the objects: the [only uses I can find](https://en.wikipedia.org/w/index.php?search=insource%3A%2FMorebits.wikipedia%2F&title=Special:Search&profile=advanced&fulltext=1&advancedSearch-current=%7B%22namespaces%22%3A%5B2%2C4%2C8%5D%7D&ns2=1&ns4=1&ns8=1&searchToken=5154a9r73jnddbqfuhgwar7d5) are old, unused copies of old Twinkle code, nothing relying on current gadgets.  `Morebits.wikipedia.namespacesFriendly` seems particularly unlikely to be widely used outside en.wiki; it removes -1 and -2 and uses `Wikipedia` for 4 and `(Article)` for 0, which we can deal with (using `(Main)` should be fine across projects).

Dunno about you @atlight @MusikAnimal but I don't think we need them?  We could keep `Morebits.wikipedia.namespaces` for the sake of backward compatibility(?), I suppose, and just copy `wgFormattedNamespaces` into it; I certainly think we'd be fine just removing `namespacesFriendly` altogether.

It's a small step, but should lay the groundwork for the above request.  I think there's a little bit in #485 that does this sort of thing, actually.

**Comment by siddharthvp** [comment]
I don't think `Morebits.wikipedia` is being used outside Twinkle at all, as
the search results show from (apart from outdated copies of Twinkle code).
It's only the popular classes like simpleWindow, wiki.page and probably
quickform that get used externally.

I suggest `Morebits.wikipedia` be removed completely.

**Comment by siddharthvp** [comment]
Regarding the issue at hand here: based on the code given here (which also appears in @Siddhartha-Ghai's [TWG](https://github.com/Siddhartha-Ghai/twinkle/commit/98ffd2baf716f17aa187528e3721dc83e0972717)), it is easy to derive a generalised function for giving namespace name regexes for any namespace: 

```javascript
function namespaceRegex(namespaceNumber) {
	var namespaceRegex = "";
	for ( var alias in mw.config.get('wgNamespaceIds') ) {
		if ( mw.config.get('wgNamespaceIds')[alias] === namespaceNumber ) {
			if (alias[0].toUpperCase() === alias[0].toLowerCase()) {
				namespaceRegex += alias;
			} else {
				namespaceRegex += '[' + alias[0].toUpperCase() + alias[0] + ']' + alias.slice(1);
			}
			namespaceRegex += '|';
		}
	}
	namespaceRegex = namespaceRegex.slice(0,-1).replace(/_/g,'[ _]');
	return namespaceRegex;
}
```
(The `if (alias[0].toUpperCase() === alias[0].toLowerCase()` logic is added in interest of non-latin script languages which don't have upper/lower case characters.)

Regarding @atlight's concern that on-the-fly regex creation logic slows the code, the solution is that we pre-compute these regexes (for the required namespaces) in twinkle.js file and store them in variables, for ready access. eg:
```javascript
Twinkle.file_ns_rgx = namespaceRegex(6);
```

Regarding whether these changes should be made here is questionable, but this must be there in any attempted internationalized version of Twinkle.

**Comment by siddharthvp** [comment]
Resolved in #1262 by @Amorymeltzer. On-the-fly regex creation is being used; don't think the 2012-era concerns of it slowing down the code are applicable any longer as today's JS engines are just so fast.
